mqtt/mqtt-bridge/src/pump/mod.rs

```diff
@@ -0,0 +1,314 @@
+    egress: Egress,
+    ingress: Ingress<H>,
+}
+
+impl<H: EventHandler> Pump<H> {
+    fn new(
+        messages_send: mpsc::Sender<PumpMessage>,
+        messages_recv: mpsc::Receiver<PumpMessage>,
+        client: MqttClient<H>,
+        subscriptions: Vec<String>,
+        loader: MessageLoader<WakingMemoryStore>,
+        store: PublicationStore<WakingMemoryStore>,
+    ) -> Result<Self, BridgeError> {
+        let client_shutdown = client.shutdown_handle()?;
+        let publish_handle = client
+            .publish_handle()
+            .map_err(BridgeError::PublishHandle)?;
+
+        let egress = Egress::new(publish_handle, loader, store);
+        let ingress = Ingress::new(client, client_shutdown);
+
+        let handle = PumpHandle::new(messages_send.clone());
+        let messages = MessagesProcessor::new(messages_recv, handle);
+
+        Ok(Self {
+            subscriptions,
+            messages_send,
+            messages,
+            egress,
+            ingress,
+        })
+    }
+
+    pub fn handle(&self) -> PumpHandle {
+        PumpHandle::new(self.messages_send.clone())
+    }
+
+    pub async fn subscribe(&mut self) -> Result<(), BridgeError> {
+        self.ingress
+            .client
+            .subscribe(&self.subscriptions) //TODO react on PumpMessage instead
+            .await
+            .map_err(BridgeError::Subscribe)?;
+
+        Ok(())
+    }
+
+    pub async fn run(mut self) {
+        info!("starting pump");
+
+        let shutdown_egress = self.egress.handle();
+        let shutdown_ingress = self.ingress.handle();
+        let shutdown_messages = self.messages.handle();
+
+        select! {
+            _= self.egress.run().fuse() => {
+                error!("egress stopped unexpectedly");
+                shutdown_ingress.shutdown().await;
+                shutdown_messages.shutdown().await;
+            },
+            _= self.ingress.run().fuse() => {
+                error!("ingress stopped unexpectedly");
+                shutdown_egress.shutdown().await;
+                shutdown_messages.shutdown().await;
+            },
+            _= self.messages.run().fuse() => {
+                info!("stopping pump");
+                shutdown_ingress.shutdown().await;
+                shutdown_egress.shutdown().await;
+            }
+        };
+
+        info!("stopped pump");
+    }
+}
+
+struct MessagesProcessorShutdownHandle(Option<PumpHandle>);
+
+impl MessagesProcessorShutdownHandle {
+    async fn shutdown(mut self) {
+        if let Some(mut sender) = self.0.take() {
+            if let Err(e) = sender.send(PumpMessage::Shutdown).await {
+                error!("unable to request shutdown for message processor. {}", e);
+            }
+        }
+    }
+}
+
+struct MessagesProcessor {
+    messages: mpsc::Receiver<PumpMessage>,
+    handle: Option<PumpHandle>,
+}
+
+impl MessagesProcessor {
+    fn new(messages: mpsc::Receiver<PumpMessage>, handle: PumpHandle) -> Self {
```

Comment:
messages processor needs a publish handle, at least for connectivity it needs to be able to send a message to the broker

Reply:
correct, I added the PumpMessageHandler trait. So you need to pass **another** PumpHandle in order this to work